File: protocols/proven-appserver/ffi/zig/src/appserver.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// appserver.zig -- Zig FFI implementation of proven-appserver.
//
// Implements the application server state machine with:
//   - 64-slot mutex-protected server pool
//   - Handler registration per server (max 64 handlers)
//   - Per-handler request type configuration
//   - Kubernetes-style health check probes (liveness/readiness/startup)
//   - Full lifecycle: Initializing -> Starting -> Running -> Draining -> Stopping -> Stopped
//   - Deploy strategy stored per instance
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching AppserverABI.Types.idr exactly.

const std = @import("std");

// =========================================================================
// Enums (matching AppserverABI.Types.idr tag assignments)
// =========================================================================

/// Request types handled by the application server (ABI tags 0-3).
pub const RequestType = enum(u8) {
    http = 0,
    websocket = 1,
    grpc = 2,
    graphql = 3,
};

/// Lifecycle states of the application server (ABI tags 0-5).
pub const LifecycleState = enum(u8) {
    initializing = 0,
    starting = 1,
    running = 2,
    draining = 3,
    stopping = 4,
    stopped = 5,
};

/// Kubernetes-style health check probes (ABI tags 0-2).
pub const HealthCheck = enum(u8) {
    liveness = 0,
    readiness = 1,
    startup = 2,
};

/// Deployment strategies (ABI tags 0-3).
pub const DeployStrategy = enum(u8) {
    rolling_update = 0,
    blue_green = 1,
    canary = 2,
    recreate = 3,
};

/// Error categories (ABI tags 0-4).
pub const ErrorCategory = enum(u8) {
    client_error = 0,
    server_error = 1,
    timeout = 2,
    circuit_open = 3,
    rate_limited = 4,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent servers.
const MAX_SERVERS: usize = 64;

/// Maximum handlers per server.
const MAX_HANDLERS: usize = 64;

/// Maximum path length in bytes.
const MAX_PATH_LEN: usize = 256;

/// A registered request handler.
const Handler = struct {
    /// URI path pattern.
    path: [MAX_PATH_LEN]u8,
    path_len: u32,
    /// Request type this handler accepts.
    req_type: RequestType,
    /// Whether this handler slot is active.
    active: bool,
};

/// Default (empty) handler.
const empty_handler: Handler = .{
    .path = [_]u8{0} ** MAX_PATH_LEN,
    .path_len = 0,
    .req_type = .http,
    .active = false,
};

/// An application server instance.
const Server = struct {
    /// Current lifecycle state.
    state: LifecycleState,
    /// Bound port.
    port: u16,
    /// Deployment strategy.
    strategy: DeployStrategy,
    /// Registered handlers.
    handlers: [MAX_HANDLERS]Handler,
    /// Number of active handlers.
    handler_count: u32,
    /// Total requests handled (monotonic counter).
    requests_handled: u64,
    /// Whether this server slot is in use.
    active: bool,
};

/// Default (empty) server.
const empty_server: Server = .{
    .state = .initializing,
    .port = 0,
    .strategy = .rolling_update,
    .handlers = [_]Handler{empty_handler} ** MAX_HANDLERS,
    .handler_count = 0,
    .requests_handled = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var servers: [MAX_SERVERS]Server = [_]Server{empty_server} ** MAX_SERVERS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SERVERS) return null;
    const idx: usize = @intCast(slot);
    if (!servers[idx].active) return null;
    return idx;
}

/// Find a handler by path within a server.
fn findHandler(idx: usize, path: []const u8) ?usize {
    for (&servers[idx].handlers, 0..) |*h, i| {
        if (h.active and h.path_len == path.len and
            std.mem.eql(u8, h.path[0..h.path_len], path))
        {
            return i;
        }
    }
    return null;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

// -- ABI version --------------------------------------------------------------

/// Returns the ABI version number. Must match Foreign.abiVersion in Idris2.
pub export fn appserver_abi_version() callconv(.c) u32 {
    return 1;
}

// -- Lifecycle ----------------------------------------------------------------

/// Create a new application server. Returns slot index (>=0) or -1 on failure.
/// The server starts in Initializing state.
pub export fn appserver_create(port: u16, strategy: u8) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (port == 0) return -1;
    if (strategy > 3) return -1;

    for (&servers, 0..) |*srv, i| {
        if (!srv.active) {
            srv.* = empty_server;
            srv.port = port;
            srv.strategy = @enumFromInt(strategy);
            srv.state = .initializing;
            srv.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a server, releasing its slot.
pub export fn appserver_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SERVERS) return;
    servers[@intCast(slot)] = empty_server;
}

// -- State queries ------------------------------------------------------------

/// Returns the current LifecycleState tag.
pub export fn appserver_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0; // initializing fallback
    return @intFromEnum(servers[idx].state);
}

// -- Lifecycle transitions ----------------------------------------------------

/// Start the server. Returns 0 on success, 1 on rejection.
/// Transitions: Initializing -> Starting.
pub export fn appserver_start(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (servers[idx].state != .initializing) return 1;

    servers[idx].state = .starting;
    return 0;
}

/// Mark the server as ready. Returns 0 on success, 1 on rejection.
/// Transitions: Starting -> Running.
pub export fn appserver_ready(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (servers[idx].state != .starting) return 1;

    servers[idx].state = .running;
    return 0;
}

// -- Handler registration -----------------------------------------------------

/// Register a request handler. Returns 0 on success, 1 on rejection.
/// Only allowed in Running state.
pub export fn appserver_register_handler(
    slot: c_int,
    path_ptr: [*]const u8,
    path_len: u32,
    req_type: u8,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (servers[idx].state != .running) return 1;
    if (path_len == 0 or path_len > MAX_PATH_LEN) return 1;
    if (req_type > 3) return 1;

    const path = path_ptr[0..path_len];

    // Reject duplicate path
    if (findHandler(idx, path) != null) return 1;

    // Find free handler slot
    for (&servers[idx].handlers) |*h| {
        if (!h.active) {
            @memcpy(h.path[0..path_len], path);
            h.path_len = path_len;
            h.req_type = @enumFromInt(req_type);
            h.active = true;
            servers[idx].handler_count += 1;
            return 0;
        }
    }
    return 1;
}

/// Returns the number of registered handlers.
pub export fn appserver_handler_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return servers[idx].handler_count;
}

// -- Request handling ---------------------------------------------------------

/// Handle an incoming request. Returns 255 on success, or ErrorCategory tag.
pub export fn appserver_handle_request(
    slot: c_int,
    path_ptr: [*]const u8,
    path_len: u32,
    req_type: u8,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return @intFromEnum(ErrorCategory.server_error);
    if (servers[idx].state != .running) {
        return @intFromEnum(ErrorCategory.server_error);
    }
    if (path_len == 0 or path_len > MAX_PATH_LEN) return @intFromEnum(ErrorCategory.client_error);
    if (req_type > 3) return @intFromEnum(ErrorCategory.client_error);

    const path = path_ptr[0..path_len];
    const hi = findHandler(idx, path) orelse return @intFromEnum(ErrorCategory.client_error);

    // Check request type match
    const handler_type = servers[idx].handlers[hi].req_type;
    const request_type: RequestType = @enumFromInt(req_type);
    if (handler_type != request_type) {
        return @intFromEnum(ErrorCategory.client_error);
    }

    servers[idx].requests_handled += 1;
    return 255; // success sentinel
}

// -- Health checks ------------------------------------------------------------

/// Check health probe. Returns 1 if healthy, 0 if not.
/// Liveness: true if not Stopped. Readiness: true if Running. Startup: true if past Initializing.
pub export fn appserver_health_check(slot: c_int, probe: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 0;
    if (probe > 2) return 0;

    const state = servers[idx].state;
    const check: HealthCheck = @enumFromInt(probe);

    return switch (check) {
        .liveness => if (state != .stopped) 1 else 0,
        .readiness => if (state == .running) 1 else 0,
        .startup => if (state != .initializing) 1 else 0,
    };
}

// -- Shutdown lifecycle -------------------------------------------------------

/// Begin draining. Returns 0 on success, 1 on rejection.
/// Transitions: Running -> Draining.
pub export fn appserver_drain(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (servers[idx].state != .running) return 1;

    servers[idx].state = .draining;
    return 0;
}

/// Stop the server. Returns 0 on success, 1 on rejection.
/// Transitions: Draining -> Stopping.
pub export fn appserver_stop(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (servers[idx].state != .draining) return 1;

    servers[idx].state = .stopping;
    return 0;
}

/// Complete cleanup. Returns 0 on success, 1 on rejection.
/// Transitions: Stopping -> Stopped.
pub export fn appserver_cleanup(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (servers[idx].state != .stopping) return 1;

    servers[idx].state = .stopped;
    servers[idx].handlers = [_]Handler{empty_handler} ** MAX_HANDLERS;
    servers[idx].handler_count = 0;
    return 0;
}

// -- Stateless transition table -----------------------------------------------

/// Check if a server state transition is valid.
pub export fn appserver_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Initializing -> Starting
    if (from == 1 and to == 2) return 1; // Starting -> Running
    if (from == 2 and to == 3) return 1; // Running -> Draining
    if (from == 3 and to == 4) return 1; // Draining -> Stopping
    if (from == 4 and to == 5) return 1; // Stopping -> Stopped
    return 0;
}

File: bindings/rust/src/appserver.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//

//! Application Server types for the proven-servers ABI.
//!
//! Formally verified application server types.
//! Mirrors the Idris2 module `AppserverABI.Types`.
//!
//! - `RequestType` -- Request protocol types.
//! - `LifecycleState` -- Application lifecycle states.
//! - `HealthCheck` -- Health check types.
//! - `DeployStrategy` -- Deployment strategies.
//! - `ErrorCategory` -- Application error categories.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// Application Server Constants
// ===========================================================================

/// Standard application server port.
pub const APP_PORT: u16 = 8080;

// ===========================================================================
// RequestType (tags 0-3)
// ===========================================================================

/// Request protocol types.
///
/// Matches `RequestType` in `AppserverABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum RequestType {
    /// HTTP (tag 0).
    Http = 0,
    /// WebSocket (tag 1).
    WebSocket = 1,
    /// gRPC (tag 2).
    Grpc = 2,
    /// GraphQL (tag 3).
    GraphQl = 3,
}

impl RequestType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Http),
            1 => Some(Self::WebSocket),
            2 => Some(Self::Grpc),
            3 => Some(Self::GraphQl),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [RequestType; 4] = [
        Self::Http, Self::WebSocket, Self::Grpc, Self::GraphQl,
    ];
}

impl fmt::Display for RequestType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// LifecycleState (tags 0-5)
// ===========================================================================

/// Application lifecycle states.
///
/// Matches `LifecycleState` in `AppserverABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum LifecycleState {
    /// Initializing (tag 0).
    Initializing = 0,
    /// Starting (tag 1).
    Starting = 1,
    /// Running (tag 2).
    Running = 2,
    /// Draining (tag 3).
    Draining = 3,
    /// Stopping (tag 4).
    Stopping = 4,
    /// Stopped (tag 5).
    Stopped = 5,
}

impl LifecycleState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Initializing),
            1 => Some(Self::Starting),
            2 => Some(Self::Running),
            3 => Some(Self::Draining),
            4 => Some(Self::Stopping),
            5 => Some(Self::Stopped),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// Whether the server is ready to handle requests.
    pub fn is_ready(self) -> bool {
        matches!(self, Self::Running)
    }

    /// All variants of this type.
    pub const ALL: [LifecycleState; 6] = [
        Self::Initializing, Self::Starting, Self::Running, Self::Draining, Self::Stopping, Self::Stopped,
    ];
}

impl fmt::Display for LifecycleState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// HealthCheck (tags 0-2)
// ===========================================================================

/// Health check types.
///
/// Matches `HealthCheck` in `AppserverABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum HealthCheck {
    /// Liveness (tag 0).
    Liveness = 0,
    /// Readiness (tag 1).
    Readiness = 1,
    /// Startup (tag 2).
    Startup = 2,
}

impl HealthCheck {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Liveness),
            1 => Some(Self::Readiness),
            2 => Some(Self::Startup),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [HealthCheck; 3] = [
        Self::Liveness, Self::Readiness, Self::Startup,
    ];
}

impl fmt::Display for HealthCheck {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// DeployStrategy (tags 0-3)
// ===========================================================================

/// Deployment strategies.
///
/// Matches `DeployStrategy` in `AppserverABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum DeployStrategy {
    /// RollingUpdate (tag 0).
    RollingUpdate = 0,
    /// BlueGreen (tag 1).
    BlueGreen = 1,
    /// Canary (tag 2).
    Canary = 2,
    /// Recreate (tag 3).
    Recreate = 3,
}

impl DeployStrategy {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::RollingUpdate),
            1 => Some(Self::BlueGreen),
            2 => Some(Self::Canary),
            3 => Some(Self::Recreate),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [DeployStrategy; 4] = [
        Self::RollingUpdate, Self::BlueGreen, Self::Canary, Self::Recreate,
    ];
}

impl fmt::Display for DeployStrategy {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// ErrorCategory (tags 0-4)
// ===========================================================================

/// Application error categories.
///
/// Matches `ErrorCategory` in `AppserverABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum ErrorCategory {
    /// ClientError (tag 0).
    ClientError = 0,
    /// ServerError (tag 1).
    ServerError = 1,
    /// Timeout (tag 2).
    Timeout = 2,
    /// CircuitOpen (tag 3).
    CircuitOpen = 3,
    /// RateLimited (tag 4).
    RateLimited = 4,
}

impl ErrorCategory {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::ClientError),
            1 => Some(Self::ServerError),
            2 => Some(Self::Timeout),
            3 => Some(Self::CircuitOpen),
            4 => Some(Self::RateLimited),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [ErrorCategory; 5] = [
        Self::ClientError, Self::ServerError, Self::Timeout, Self::CircuitOpen, Self::RateLimited,
    ];
}

impl fmt::Display for ErrorCategory {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn request_type_roundtrip() {
        for v in RequestType::ALL {
            let tag = v.to_tag();
            let decoded = RequestType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(RequestType::from_tag(4).is_none());
    }

    #[test]
    fn lifecycle_state_roundtrip() {
        for v in LifecycleState::ALL {
            let tag = v.to_tag();
            let decoded = LifecycleState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(LifecycleState::from_tag(6).is_none());
    }

    #[test]
    fn health_check_roundtrip() {
        for v in HealthCheck::ALL {
            let tag = v.to_tag();
            let decoded = HealthCheck::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(HealthCheck::from_tag(3).is_none());
    }

    #[test]
    fn deploy_strategy_roundtrip() {
        for v in DeployStrategy::ALL {
            let tag = v.to_tag();
            let decoded = DeployStrategy::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(DeployStrategy::from_tag(4).is_none());
    }

    #[test]
    fn error_category_roundtrip() {
        for v in ErrorCategory::ALL {
            let tag = v.to_tag();
            let decoded = ErrorCategory::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(ErrorCategory::from_tag(5).is_none());
    }

    #[test]
    fn constants_match_idris() {
        assert_eq!(APP_PORT, 8080);
    }

}
